process-util: handle pidfd_spawn() returning E2BIG

kkourt opened the pull request:
process-util: handle pidfd_spawn() returning E2BIG

In some kernels (specifically, 5.4) even though the clone3 syscall is
supported, setting CLONE_INTO_CGROUP is not. The error message returned
in this case is E2BIG.

If posix_spawn_wrapper encounters this error, it does not retry, and
cannot spawn any programs in said kernels.

This commit adds a check for the E2BIG error and retries pidfd_spawn()
without the POSIX_SPAWN_SETCGROUP flag.

Originally, this issue was reported in https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1077204.

Fixes #33613

kkourt commented:
CC: @bluca who introduced pidfd_spawn in 2e106312e2e8e8f88516039446abcf7d9afdaff2

bluca commented on src/basic/process-util.c:
```diff
@@ -2107,7 +2107,7 @@ int posix_spawn_wrapper(
                 return FLAGS_SET(flags, POSIX_SPAWN_SETCGROUP);
         }
-        if (ERRNO_IS_NOT_SUPPORTED(r)) {
+        if (ERRNO_IS_NOT_SUPPORTED(r) || r == E2BIG) {
```
If this is only an issue due to POSIX_SPAWN_SETCGROUP, then instead of falling back to posix_spawn, please add another branch just above the r == 0 that retries removes POSIX_SPAWN_SETCGROUP from flags and retries pidfd_spawn

kkourt replied:
Thanks, pushed a new version with this change.

YHNdnzj commented:
> In some kernels (specifically, 5.4) even though the clone3 syscall is supported, setting CLONE_INTO_CGROUP is not. The error message returned in this case is E2BIG.

Hmm, but this is neither documented nor sane behavior, no? Has this been raised to kernel/glibc folks first?

YHNdnzj commented:
Hmm, https://lore.kernel.org/lkml/20240808-clone3-shadow-stack-v8-8-0acf37caf14c@kernel.org/ suggests that this is indeed considered valid for clone3(). I wonder glibc should document this properly or convert the errno to e.g. EINVAL. But this patch makes sense nonetheless then.

yuwata commented on src/basic/process-util.c:
```diff
@@ -2107,7 +2107,7 @@ int posix_spawn_wrapper(
                 return FLAGS_SET(flags, POSIX_SPAWN_SETCGROUP);
         }
-        if (ERRNO_IS_NOT_SUPPORTED(r)) {
+        if (ERRNO_IS_NOT_SUPPORTED(r) || r == E2BIG) {
```
Please add a short comment what `E2BIG` means here. Not only in the commit message.

kkourt replied:
Thanks, pushed a new version that includes a comment as well.

K900 commented:
I'm pretty sure this is actually a glibc bug - it's failing because it's setting `clone_args.cgroup` unconditionally even on kernels where it's not available, instead of taking the fallback.

poettering commented:
I think kernel/glibc behaviour does make some sense, and yes, we should handle this gracefully, i.e. try to disable the POSIX_SPAWN_SETCGROUP flag, then try again, and if it works now, cache this fact.

poettering commented on src/basic/process-util.c:
```diff
@@ -2100,6 +2100,15 @@ int posix_spawn_wrapper(
         _cleanup_close_ int pidfd = -EBADF;
 
         r = pidfd_spawn(&pidfd, path, NULL, &attr, argv, envp);
+        if (r == E2BIG) {
```
please also check if POSIX_SPAWN_SETCGROUP was actually set here. otherwise E2BIG might mean something unrelated after all.

poettering commented on src/basic/process-util.c:
```diff
@@ -2100,6 +2100,15 @@ int posix_spawn_wrapper(
+                if (r != 0)
+                        return -r;
+                r = pidfd_spawn(&pidfd, path, NULL, &attr, argv, envp);
+        }
```
and please, as suggested earlier, cache the result, i.e. add "static bool setcgroup_supported = true;" somewhere as a local static variable here, and then set it to false if we realize it doesn't actually work. Then, always check this field before we try to do anything.

kkourt replied:
OK, did so in a second patch. I only update the static variable if we were successful because otherwise, there might be something else going wrong.

YHNdnzj commented on src/basic/process-util.c:
```diff
@@ -2117,11 +2126,14 @@ int posix_spawn_wrapper(
-                return -r;
+         * flags to remove the cgroup one, which is what redirects to clone3().
+         * Clear POSIX_SPAWN_SETCGROUP if it was not already done. */
+        if (flags & POSIX_SPAWN_SETCGROUP) {
```
FLAGS_SET

yuwata commented:
@kkourt Could you squash the two commits? I do not think the change should be split into two.

kkourt commented:
> @kkourt Could you squash the two commits? I do not think the change should be split into two.

Sounds good. Squashed the commits and moved the declaration of the boolean up requested.

YHNdnzj commented on src/basic/process-util.c:
```diff
@@ -2099,7 +2100,28 @@ int posix_spawn_wrapper(
         _cleanup_close_ int pidfd = -EBADF;
 
+        /* If we have observed that SPAWN_CGROUP is not supported, remove the flag */
+        if (!setcgroup_supported && FLAGS_SET(flags, POSIX_SPAWN_SETCGROUP)) {
```
Please just skip the opening of cgroup altogether if we know in advance it would be ignored anyway.

yuwata replied:
Yeah, please see the above code in my comment.

poettering replied:
yes, the opening of the cgorup is what is slow, and if it doesn't work we really shouldn't bother.

poettering commented on src/basic/process-util.c:
```diff
@@ -2099,7 +2100,28 @@ int posix_spawn_wrapper(
+                if (r != 0)
+                        return -r;
+                r = pidfd_spawn(&pidfd, path, NULL, &attr, argv, envp);
+                /* if pidfd_spawn was succesfull after removing SPAWN_CGROUP,
```
successfull → successful

kkourt commented:
Thanks for the reviews, I believe all comments were addressed in the latest push.

poettering commented on src/basic/process-util.c:
```diff
@@ -2099,7 +2100,28 @@ int posix_spawn_wrapper(
+                r = posix_spawnattr_setflags(&attr, flags);
+                if (r != 0)
+                        return -r;
+        }
```
you can drop this whole block now

kkourt replied:
ah yes, indeed.

I left the check in:
```
+        if (FLAGS_SET(flags, POSIX_SPAWN_SETCGROUP)) {
+                flags &= ~POSIX_SPAWN_SETCGROUP;
+                r = posix_spawnattr_setflags(&attr, flags);
+                if (r != 0)
+                        return -r;
+        }
```

But maybe it can be removed as well?

yuwata replied:
The change should be OK. Let's keep it.

yuwata reviewed (approved):
LGTM.
